constraints/miden-vm/hash.air

```diff
@@ -0,0 +1,156 @@
+
+    # Set to 1 when either f_hout = 1 or f_sout = 1 in the current row.
+    let f_out = k0 * binary_not(s[0]) * binary_not(s[1])
+
+    # Set to 1 when selector flags are (1,0,1) on rows which are 1 less than a multiple of 8.
+    let f_mpa = k0 * s[0] * binary_not(s[1]) * s[2]
+    # Set to 1 when selector flags are (1,1,0) on rows which are 1 less than a multiple of 8.
+    let f_mva = k0 * s[0] * s[1] * binary_not(s[2])
+    # Set to 1 when selector flags are (1,1,1) on rows which are 1 less than a multiple of 8.
+    let f_mua = k0 * s[0] * s[1] * s[2]
+
+    # b is the value of the bit which is discarded during shift by one bit to the right.
+    let b = i - 2 * i'
+
+    # Flag that allows to enforce constraint that b is binary only when a new node is absorbed into
+    # the hasher state
+    let f_an = f_mp + f_mv + f_mu + f_mpa + f_mva + f_mua
+
+    # Enforce that b is binary only when a new node is absorbed into the hasher state.
+    enf f_an * (b^2 - b) = 0
+
+    # Enforce that when a computation is finished i = 0.
+    enf f_out * i = 0
+
+    # Enforce that the value in i is copied over to the next row unless we are absorbing a new row 
+    # or the computation is finished.
+    let absorbing_or_comp_finished = 1 - f_an - f_out
+    enf is_copied([i]) when absorbing_or_comp_finished
+
+
+# Enforce hasher state constraints
+ev hasher_state(main: [s[3], h[12], i]):
+    let f_mp = k2 * s[0] * binary_not(s[1]) * s[2]
+    # Set to 1 when selector flags are (1,1,0) on rows which are multiples of 8.
+    let f_mv = k2 * s[0] * s[1] * binary_not(s[2])
+    # Set to 1 when selector flags are (1,1,1) on rows which are multiples of 8.
+    let f_mu = k2 * s[0] * s[1] * s[2]
+
+    let f_abs_node = f_mp + f_mv + f_mu
+
+    # Set to 1 when selector flags are (1,0,0) on rows which are 1 less than a multiple of 8.
+    let f_abp = k0 * s[0] * binary_not(s[1]) * binary_not(s[2])
+
+    # b is the value of the bit which is discarded during shift by one bit to the right.
+    let b = i - 2 * i'
+
+    # Enforce that when absorbing the next set of elements into the state during linear hash 
+    # computation (i.e. f_abp = 1) the first 4 elements (the capacity portion) are carried over to 
+    # the next row.
+    enf f_abp * (h[j]' - h[j]) = 0 for j in 0..4
+
+    # Enforce that when absorbing the next node during Merkle path computation 
+    # (i.e. f_mp + f_mv + f_mu = 1), the result of the previous hash (h[4], ..., h[7]) are copied 
+    # over either to (h[4]', ..., h[7]') or to (h[8]', ..., h[11]') depending on the value of b.
+    # TODO: once the support for combination of constraint comprehension and selectors is 
+    # implemented, update this constraint as:
+    # ```
+    # match enf:
+    #     is_copied(h[j + 4]) for j in 0..4 when !b & f_abs_node
+    #     h[j + 8]' = h[j + 4] for j in 0..4 when b & f_abs_node
+    # ```
+    enf f_abs_node * ((1 - b) * is_copied(h[j + 4]) + b * (h[j + 8]' - h[j + 4])) = 0 for j in 0..4
```

Comment:
I don't think we need to wait, since this isn't going into the tests. I would just change this now:

 ```
 match enf:
     is_copied(h[j + 4]) for j in 0..4 when !b & f_abs_node
     h[j + 8]' = h[j + 4] for j in 0..4 when b & f_abs_node
 ```